### PR TITLE
Add basic setup switching for the tournament client

### PR DESCRIPTION
**Pull request by MiraiSubject**
- [x] Depends on https://github.com/ppy/osu/pull/9247 for tests to pass

From https://github.com/ppy/osu/issues/9199, this fulfils 
> Create a default tournament profile and the necessary migrations for the flags, mods, videos and other assets.

This PR changes the default path for tournaments from `%APPDATA%\osu\tournament`  to `%APPDATA%\osu\tournaments\default`.
Migrations for this are implemented so that this change does not break existing configurations and the necessary tests are also written for that. The migration also includes moving the `bracket.json` to the same directory. 

Upon migration a `tournament.ini` gets created with the following default value:
```
CurrentTournament = default
```
## Post-migration scenario
The user can now add directories in the following format:
```
(%APPDATA%/osu/tournaments)
tournament-name
 ┃ ┣ config
 ┃ ┃ ┗ drawings.ini
 ┃ ┣ flags
 ┃ ┣ mods
 ┃ ┣ videos
 ┃ ┣ bracket.json
 ┃ ┣ drawings.txt
 ┃ ┗ drawings_results.txt
```
and then change the `tournament.ini` to:
```
CurrentTournament = tournament-name
```
Because of this, switching the storage requires a reboot of the tournament client.

### CODE REVIEW

**Comment by peppy** [comment]
Generally if something depends on another branch/PR, you should rebase it on top of that (so it will compile and work for testing).

**Comment by MiraiSubject** [comment]
> Generally if something depends on another branch/PR, you should rebase it on top of that (so it will compile and work for testing).

I locally tested by merging this branch into https://github.com/ppy/osu/pull/9247, stashed any changes I wanted to keep and then applied it to this branch. It's a very roundabout way of doing things, but it was the only way I was familiar and comfortable with. 

I'll look into rebasing.

**Comment by peppy** [comment]
It's fine this time around, just a heads-up for future contributions.

File: osu.Game.Tournament/IO/TournamentStorage.cs
Comment (peppy) [inline osu.Game.Tournament/IO/TournamentStorage.cs:22]: Why can't this just take in a storage, rather than the host itself?

File: osu.Game.Tournament/IO/TournamentStorage.cs
Comment (peppy) [inline osu.Game.Tournament/IO/TournamentStorage.cs:61]: please learn how to avoid doing this by using functions.

File: osu.Game.Tournament/IO/TournamentStorage.cs
Comment (peppy) [inline osu.Game.Tournament/IO/TournamentStorage.cs:68]: why can't you use the existing `Migrate` method? this is mostly copy-paste of existing implementation.

File: osu.Game.Tournament/IO/TournamentStorage.cs
Comment (peppy) [inline osu.Game.Tournament/IO/TournamentStorage.cs:97]: don't copy paste please.

File: osu.Game.Tournament/IO/TournamentStorage.cs
Comment (peppy) [inline osu.Game.Tournament/IO/TournamentStorage.cs:20]: why even bother? isn't there only one config file ever?

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by MiraiSubject** [comment]
There a couple of changes I made to this: 

- Introduce a new abstract class `MigratableStorage` that inherits from `WrappedStorage`
- Change the inheritance for `OsuStorage` and `TournamentStorage` to `MigratableStorage`
- `TournamentStorage` now takes in an existing storage as ctor parameter
- File moving in `TournamentStorage` is now called `moveFileIfExists` to prevent filling up `migrate()` up with if-statements.
- `drawings.ini` has been moved back from `tournaments/default/config` to `tournaments/default`

`MigratableStorage` now contains the Copy, Delete and Attempt operation methods as well as the constants for `ignore_files` and `ignore_directories`

I also made sure all the tests for `CustomDataDirectoryTest` still pass. 

I'm not sure if the `ignore_files` and `ignore_directories` should stay there, because imo every `MigratableStorage` should have their own array of ignored files and directories and now it's globally declared across every subclass. I just haven't found a solution to this yet that does not look bad. If anybody has an idea about this, let me know.

File: osu.Game/IO/MigratableStorage.cs
Comment (bdach) [inline osu.Game/IO/MigratableStorage.cs:23]: The obvious way to make this customisable by consumers is by making these into protected get-only virtual properties.

File: osu.Game.Tournament/Components/TourneyVideo.cs
Comment (bdach) [inline osu.Game.Tournament/Components/TourneyVideo.cs:33]: This string interpolation is not required any more.
```suggestion
            var stream = storage.VideoStore.GetStream(filename);
```

File: osu.Game.Tournament/Configuration/TournamentStorageManager.cs
Comment (bdach) [inline osu.Game.Tournament/Configuration/TournamentStorageManager.cs:21]: Is this even necessary if it's going to be overwritten upon first execution of `TournamentStorage` anyway?

File: osu.Game.Tournament/IO/TournamentStorage.cs
Comment (bdach) [inline osu.Game.Tournament/IO/TournamentStorage.cs:15]: The `internal` looks really strange to me. I'd write this as

```suggestion
        public readonly TournamentVideoResourceStore VideoStore { get; }
```

File: osu.Game.Tournament/Screens/Drawings/Components/StorageBackedTeamList.cs
Comment (bdach) [inline osu.Game.Tournament/Screens/Drawings/Components/StorageBackedTeamList.cs:19]: Is this required? A `TournamentStorage` is a `Storage` still...

File: osu.Game.Tournament/Screens/Drawings/DrawingsScreen.cs
Comment (bdach) [inline osu.Game.Tournament/Screens/Drawings/DrawingsScreen.cs:44]: This one is confusing too, but I can understand why this would be required more (due to caching in `TournamentGameBase`, I presume). You can `CacheAs<Storage>`, fwiw.

File: osu.Game.Tests/NonVisual/CustomDataDirectoryTest.cs
Comment (MiraiSubject) [inline osu.Game.Tests/NonVisual/CustomDataDirectoryTest.cs:161]: Not sure if this is the right way to prevent a Nullref exception and it also looks weird to loop over an empty array

Comment (peppy) [reply osu.Game.Tests/NonVisual/CustomDataDirectoryTest.cs:161]: You're correct in thinking this is not good code. just do a null check *before* iterating instead (you can use it to surround both the two `foreach` statements).

File: osu.Game.Tests/NonVisual/CustomDataDirectoryTest.cs
Comment (bdach) [inline osu.Game.Tests/NonVisual/CustomDataDirectoryTest.cs:155]: Shouldn't this be an actual assertion, as in the following?

```csharp
Assert.That(osuStorage, Is.Not.Null);
```

Imagine this regresses for some reason and `osuStorage` becomes null; this `if` will then mean that the test will silently skip over the nested asserts and stop testing if the ignore lists work at all.

(As an aside, I think it'd also make sense to change type of `osuStorage` to `MigratableStorage` as it's the most general type possible in this scenario)

File: osu.Game/IO/MigratableStorage.cs
Comment (bdach) [inline osu.Game/IO/MigratableStorage.cs:18]: This is a weird modifier mix. Any consumers from a different assembly that might want to override this will not be able to. I think this calls for the [rarely-seen](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/protected-internal)

```suggestion
        protected internal virtual string[] IgnoreDirectories => Array.Empty<string>();
        protected internal virtual string[] IgnoreFiles => Array.Empty<string>();
```

(Or heck, just make it public, I don't think that's that much of a deal.)

Needs to be fixed up in the override too if this happens.

File: osu.Game.Tests/NonVisual/CustomDataDirectoryTest.cs
Comment (MiraiSubject) [reply osu.Game.Tests/NonVisual/CustomDataDirectoryTest.cs]: Good point! I'll add that and change the cast right away. I'll also include the property accessor change right away.

File: osu.Game/IO/MigratableStorage.cs
Comment (MiraiSubject) [reply osu.Game/IO/MigratableStorage.cs]: I've changed it to public for now, will await further feedback about this in case `protected` becomes the preferred way to go.

**Comment by MiraiSubject** [comment]
I've been looking into making `Migrate` more sane.

Right now I moved the original `Migrate` from `OsuStorage` to `MigratableStorage` (+xmldoc) so that consumers can use this as a general purpose migrator.

I've also looked into making the `TournamentStorage`'s `Migrate` use `MigratableStorage`'s but that's not possible because the destination is nested inside the source. That goes along with the structure of what's being migrated is different from what it would be in a general case. 

Before:
```
(%APPDATA%/osu/)
... cache, files, etc.
 ┃ ┣ tournament/
 ┃ ┃ ┣  flags
 ┃ ┃ ┣  mods
 ┃ ┃ ┗  videos
 ┃ ┣  bracket.json
 ┃ ┣ drawings.txt.
 ┃ ┣ drawings_results.txt.
 ┃ ┗ drawings.ini
...  framework.ini, online.db, etc.
```

After: 
```
(%APPDATA%/osu/tournaments)
default
 ┃ ┣ flags
 ┃ ┣ mods
 ┃ ┣ videos
 ┃ ┣ bracket.json
 ┃ ┣ drawings.txt
 ┃ ┣ drawings.ini
 ┃ ┗ drawings_results.txt
```

**Comment by bdach** [comment]
As a heads up, but this is probably going to have major merge conflicts with #9417 / #9418...

As for the issue mentioned I'd probably advise that the `Migrate()` method be overridable and that `TournamentStorage` hoist the files one level out of `tournament/` as a post-migration step.

**Comment by MiraiSubject** [comment]
Alright thanks! I'll hold off from any more big changes until after those are merged

I made `Migrate` already overridable and its implementation in `TournamentStorage` is being completely overridden without ever calling `base` because this exception in `MigratableStorage` will get thrown:

```csharp
if (sourceUri.IsBaseOf(destinationUri))
    throw new ArgumentException("Destination provided is inside the source", destination.FullName);
```

I also believe at the same time that `TournamentStorage`'s `Migrate` is different enough to not need to call `base`.

I could just be talking about some non-issue and that overriding like this is fine, but I am not sure. Hopefully it will come to light once it can be reviewed again.

**Comment by MiraiSubject** [comment]
I resolved the conflicts in an incoming commit and reran the tests locally again to make sure they didn't fail. 

I thought some more about this and my comment above can actually be disregarded. The intention of the migration method is that it is supposed to be a temporary method so that existing configurations get migrated after users update the client.

For future versions users would be instructed to create their new configurations in the format of `tournaments/<tournament acronym>` and then edit the `tournament.ini` appropriately to `CurrentTournament = <tournament acronym>`. 

The end goal is to put this functionality of adding and selecting tournaments in the UI itself.

**Review by peppy — APPROVED** [review]
Should be good to go now.

**Pull request merged**
